fv3core/stencils/fxadv.py

```diff
@@ -1,314 +1,381 @@
+    return vt
 
 
-@gtstencil()
-def vt_y_edge(vc: sd, cosa_v: sd, ut: sd, vt: sd):
-    with computation(PARALLEL), interval(0, -1):
-        vt[0, 0, 0] = vc - 0.25 * cosa_v * (
-            ut[0, -1, 0] + ut[1, -1, 0] + ut[0, 0, 0] + ut[1, 0, 0]
-        )
-
+@gtscript.function
+def ut_y_edge(uc, sin_sg1, sin_sg3, ut, dt):
+    from __externals__ import i_end, i_start
 
-@gtstencil()
-def vt_x_edge(vc: sd, sin_sg2: sd, sin_sg4: sd, vt: sd, *, dt: float):
-    with computation(PARALLEL), interval(0, -1):
-        vt[0, 0, 0] = (vc / sin_sg4[0, -1, 0]) if (vc * dt > 0) else (vc / sin_sg2)
+    with horizontal(region[i_start, :], region[i_end + 1, :]):
+        ut = (uc / sin_sg3[-1, 0, 0]) if (uc * dt > 0) else (uc / sin_sg1)
+    return ut
 
 
 @gtscript.function
-def ra_x_func(area, xfx_adv):
-    return area + xfx_adv - xfx_adv[1, 0, 0]
+def ut_x_edge(uc, cosa_u, vt, ut):
+    from __externals__ import i_end, i_start, j_end, j_start, local_ie, local_is
+
+    # TODO: parallel to what done for the vt_y_edge section
+    utmp = ut
+    with horizontal(
+        region[local_is : local_ie + 2, j_start - 1 : j_start + 1],
+        region[local_is : local_ie + 2, j_end : j_end + 2],
+    ):
+        ut = uc - 0.25 * cosa_u * (vt[-1, 0, 0] + vt + vt[-1, 1, 0] + vt[0, 1, 0])
+    with horizontal(
+        region[i_start : i_start + 2, j_start - 1 : j_start + 1],
+        region[i_start : i_start + 2, j_end : j_end + 2],
+        region[i_end : i_end + 2, j_start - 1 : j_start + 1],
+        region[i_end : i_end + 2, j_end : j_end + 2],
+    ):
+        ut = utmp
+    return ut
 
 
-@gtstencil()
-def xfx_adv_stencil(
-    ut: sd,
-    rdxa: sd,
-    area: sd,
-    dy: sd,
-    sin_sg1: sd,
-    sin_sg3: sd,
-    crx_adv: sd,
-    xfx_adv: sd,
-    ra_x: sd,
-    dt: float,
-):
-    with computation(PARALLEL), interval(...):
-        prod = dt * ut
-        crx_adv = prod * rdxa[-1, 0, 0] if prod > 0 else prod * rdxa
-        xfx_adv = dy * prod * sin_sg3[-1, 0, 0] if prod > 0 else dy * prod * sin_sg1
-        ra_x = ra_x_func(area, xfx_adv)
+@gtscript.function
+def vt_y_edge(vc, cosa_v, ut, vt):
+    from __externals__ import i_end, i_start, j_end, j_start, local_je, local_js
+
+    # This works for 6 ranks, but not 54:
+    # with horizontal(region[i_start - 1: i_start + 1, j_start + 2:j_end], \
+    #                region[i_end : i_end + 2, j_start+2:j_end]):
+    #    vt = vc - 0.25 * cosa_v * (
+    #        ut[0, -1, 0] + ut[1, -1, 0] + ut + ut[1, 0, 0]
+    #    )
+    # original bounds with stencil calls
+    # j1 = grid().js + 2 if grid().south_edge else grid().js
+    # j2 = grid().je if grid().north_edge else grid().je + 2
+    # TODO: this is a hack, copying vt to vtmp to 'correct' the edges
+    # Can we *just* apply edge calculations in the correct regions without overcomputing
+    # rank 0, 1, 2: local_js + 2:local_je + 2
+    # rank 3, 4, 5: local_js:local_je + 2
+    # rank 6, 7, 8: local_js:local_je
+    vtmp = vt
```

Comment:
Same question as above...

Reply:
Yeah I left the todo because I don't know if there is a good way to not have that temporary field copied from the vt... Maybe it isn't that impactful?